### PR TITLE
Closes #16394. Fix issues in `migrate/create`

### LINKED ISSUE #16394: Moved from [yiisoft/yii2-gii#368] Incorrect result when creating migration with fields
Original issue https://github.com/yiisoft/yii2-gii/issues/368

### What steps will reproduce the problem?

yii migrate/create add_deleted_at_column_deleted_by_column_to_accountable_person_table --migrationPath=backend/modules/accountablePerson/migrations --interactive=0 --fields="deleted_at:dateTime:notNull:defaultValue(0000-00-00 00:00:00),deleted_by:integer:foreignKey(user)" --useTablePrefix=true

### What's expected?

    /**
    * Handles adding `deleted_at`, `deleted_by` to table `accountable_person`.
    */

### What do you get instead?

    /**
    * Handles adding deleted_at_column_deleted_by to table `accountable_person`.
    */

### What's expected?

    $this->addColumn('{{%accountable_person}}', 'deleted_at', $this->dateTime()->notNull()->defaultValue('0000-00-00 00:00:00'));

### What do you get instead?

    $this->addColumn('{{%accountable_person}}', 'deleted_at', $this->dateTime()->notNull()->defaultValue('0000-00-00 00()->00()->00')());

### What's expected?

    $this->createIndex(
            'idx-accountable_person-deleted_by',
    ...
    $this->addForeignKey(
            'fk-accountable_person-deleted_by',

I'm not realy shure what to expect here but whats generated does not look OK to me. Maybe it can be  'idx-{{%accountable_person}}-deleted_by' and `fk-{{%accountable_person}}-deleted_by`

### What do you get instead?

    $this->createIndex(
            '{{%idx-accountable_person-deleted_by}}',
    ...
    $this->addForeignKey(
            '{{%fk-accountable_person-deleted_by}}',

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.16-dev
| PHP version      | PHP 5.6.9

### PR DESCRIPTION
**Pull request by alexkart**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #16394

### CODE REVIEW

File: framework/console/controllers/MigrateController.php [hunk 1]
```diff
@@ -523,6 +523,35 @@ protected function parseFields()
         ];
     }
 
+    /**
+     * Splits field into chunks
+     *
+     * @param $field
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]:
```suggestion
     * @param string $field
```

File: framework/console/controllers/MigrateController.php [hunk 2]
```diff
@@ -523,6 +523,35 @@ protected function parseFields()
...
+     * @return array|array[]|false|string[]
+     */
+    protected function splitFieldIntoChunks($field)
+    {
+        $hasDoubleQuotes = false;
+        $regex = "/defaultValue\(.*?:.*?\)/";
+        preg_match_all($regex, $field, $matches);
+        if (isset($matches[0][0])) {
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: Can't `if` check be done for `preg_match_all` itself?

File: framework/console/controllers/MigrateController.php [hunk 3]
```diff
@@ -523,6 +523,35 @@ protected function parseFields()
...
+            $hasDoubleQuotes = true;
+            $origDefaultValue = $matches[0][0];
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]:
```suggestion
            $originalDefaultValue = $matches[0][0];
```

File: framework/console/controllers/MigrateController.php [hunk 4]
```diff
@@ -523,6 +523,35 @@ protected function parseFields()
...
+            $defaultValue = str_replace(':', '{{colon}}', $origDefaultValue);
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]:
```suggestion
            $defaultValue = str_replace(':', '{{colon}}', $originalDefaultValue);
```

File: framework/console/controllers/MigrateController.php [hunk 5]
```diff
@@ -523,6 +523,35 @@ protected function parseFields()
...
+    /**
+     * Splits field into chunks
+     *
+     * @param $field
+     * @return array|array[]|false|string[]
+     */
+    protected function splitFieldIntoChunks($field)
+    {
+        $hasDoubleQuotes = false;
+        $regex = "/defaultValue\(.*?:.*?\)/";
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: The variable is used only once. There's no need to introduce it.

File: framework/console/controllers/MigrateController.php [hunk 6]
```diff
@@ -523,6 +523,35 @@ protected function parseFields()
...
+        preg_match_all($regex, $field, $matches);
+        if (isset($matches[0][0])) {
+            $hasDoubleQuotes = true;
+            $origDefaultValue = $matches[0][0];
+            $defaultValue = str_replace(':', '{{colon}}', $origDefaultValue);
+            $field = str_replace($origDefaultValue, $defaultValue, $field);
+        }
+
+        $chunks = preg_split('/\s?:\s?/', $field, null);
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]:
```suggestion
        $chunks = preg_split('/\s?:\s?/', $field);
```

File: framework/console/controllers/MigrateController.php [hunk 7]
```diff
@@ -523,6 +523,35 @@ protected function parseFields()
...
+    protected function splitFieldIntoChunks($field)
+    {
+        $hasDoubleQuotes = false;
+        $regex = "/defaultValue\(.*?:.*?\)/";
+        preg_match_all($regex, $field, $matches);
+        if (isset($matches[0][0])) {
+            $hasDoubleQuotes = true;
+            $origDefaultValue = $matches[0][0];
+            $defaultValue = str_replace(':', '{{colon}}', $origDefaultValue);
+            $field = str_replace($origDefaultValue, $defaultValue, $field);
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]:
```suggestion
            $field = str_replace($originalDefaultValue, $defaultValue, $field);
```

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/console/controllers/MigrateController.php (hunk 2 above)
Comment (alexkart) [reply framework/console/controllers/MigrateController.php]: I think it's safer to check `$matches[0][0]`.

**Comment by alexkart** [comment]
Ready for review.
Case 3 from the issue seems correct to me, nothing to fix there.

**Comment by samdark** [comment]
Yes, case 3 is OK as is.

**Review by samdark — APPROVED** [review]
Looks good. Please add a line for CHANGELOG.

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**